### PR TITLE
Reimplement ExDoc.Refs.visibility helper

### PR DESCRIPTION
**Pull request by eksperimental**
It corrects the visibility of functions that start with an underscore and that no docs have been defined.

Closes https://github.com/elixir-lang/ex_doc/issues/1308

/cc @jonatanklosko

### CODE REVIEW

File: test/ex_doc/refs_test.exs
Comment (eksperimental) [inline test/ex_doc/refs_test.exs:63]: Is there anything we can do do read the @doc attributes of the docs in this module?

File: test/support/with_without_module_doc.ex
Comment (eksperimental) [inline test/support/with_without_module_doc.ex:1]: I tried different approaches and this seemed to be the easiest way to have the module avaible in the refs and the autlolink tests

File: lib/ex_doc/refs.ex
Comment (eksperimental) [inline lib/ex_doc/refs.ex:146]: This could be abstracted into a function, as it is also used in the Retriever module. But the question is where to place it. For now we just have duplicated code.

File: lib/ex_doc/refs.ex
Comment (jonatanklosko) [inline lib/ex_doc/refs.ex:139]: I believe the second condition could be simplified to `doc == %{}`.

File: test/ex_doc/autolink_test.exs
Comment (jonatanklosko) [inline test/ex_doc/autolink_test.exs:457]: `file` and `id` don't seem relevant here any reason for including these?

File: test/support/with_without_module_doc.ex
Comment (jonatanklosko) [inline test/support/with_without_module_doc.ex:44]: Relying on arity makes it necessary to look into this module to get the idea. Perhaps we could encode the information in function names like `fun_no_doc`, `fun_doc_false`, `fun_with_doc`?

File: test/ex_doc/refs_test.exs
Comment (jonatanklosko) [reply test/ex_doc/refs_test.exs:63]: The `Refs` module uses `Code.fetch_docs`, which requires the bytecode to be written to disk ([reference](https://hexdocs.pm/elixir/writing-documentation.html#code-fetch_docs-1)) and that's not the case for in-memory modules. I don't know if there's any way around that.

File: lib/ex_doc/refs.ex
Comment (jonatanklosko) [inline lib/ex_doc/refs.ex:157]: Is there any case where we would want to behave differently than `ExDoc.Retriever.doc?`? If that's not the case, then perhaps we could extract that boolean `doc?` function and then just convert it to `:public/:hidden` here?

**Review by jonatanklosko — COMMENTED** [review]
The solution looks good to me, a few tiny comments =)

File: lib/ex_doc/refs.ex
Comment (eksperimental) [reply lib/ex_doc/refs.ex]: Implemented. Thank you

File: test/ex_doc/autolink_test.exs
Comment (eksperimental) [reply test/ex_doc/autolink_test.exs]: Removed.

File: test/support/with_without_module_doc.ex
Comment (eksperimental) [reply test/support/with_without_module_doc.ex:44]: Done. this is much better now.

File: lib/ex_doc/refs.ex
Comment (eksperimental) [reply lib/ex_doc/refs.ex:157]: This is the only suggestion that I have not implemented. Do you mind sharing the implementation of what you mean?

Comment (jonatanklosko) [reply lib/ex_doc/refs.ex:157]: I looked into that and the intent of these functions is in fact different - `ExDoc.Retriever.doc?` is used only to filter which functions/macros should be shown in the doc.

I do however think it would make sense to have a single source of truth regarding which elements get documentation and which are ignored. Currently there are two separate sources:

- the `Retriever` which ignores a given module/function/macro/... based on some rules
- the `Refs` cache that keeps track of module/function/macro/... "visibility"

The question is whether these two should be in line with each other in all cases. The issue hints me that they should, but I may be missing something here.

What I had in mind is a single function that would determine if the given module or function should end up in the generated documentation or rather be ignored. This function could then be used in all the relevant places. [Here](https://github.com/eksperimental/ex_doc/compare/functions_starting_with_underscore...jonatanklosko:pr1313) I implemented such function (`should_have_doc?`) and tried to make it match the current behavior I saw in several different places. I think it removes much redundancy, but someone should confirm that this approach makes sense.

Comment (eksperimental) [reply lib/ex_doc/refs.ex:167]: @jonatanklosko  I think covering those issues in this PR it will deviate the purpose too much.
I have taken the ideas of unifying regarding to visibility and put it all in one function.
So now what this PR does is laying the foundation for the logic regarding the visibility of functions.
I think once we merge this we can look into your proposal. What do you think?

Thank you for your contributions

Comment (jonatanklosko) [reply lib/ex_doc/refs.ex:167]: Sure, what I suggested was more of a refactoring and can definitely be discussed separately. Your changes surely fix the problem.

File: test/ex_doc/autolink_test.exs
Comment (jonatanklosko) [inline test/ex_doc/autolink_test.exs:454]: Other than that, everything looks good to me ^^

Comment (eksperimental) [reply test/ex_doc/autolink_test.exs:454]: That is intentional. I did' t want the reviewing processes be swamped with unrelated changes. I would leave it in this PR, and in a next commit split it into tests.

Comment (jonatanklosko) [reply test/ex_doc/autolink_test.exs:454]: Alright, sounds good!

**Comment by eksperimental** [comment]
Guys, I am planning on going ahead and merge this if nobody has any comments.

File: lib/ex_doc/refs.ex
Comment (josevalim) [inline lib/ex_doc/refs.ex:103]: I don't see why we pass the module, we never use it. :)

File: lib/ex_doc/refs.ex
Comment (josevalim) [inline lib/ex_doc/refs.ex:107]: I think this can be simpler. Something like this:

    if module_visibility == :hidden do
      :hidden
    else
      def_visibility(ref_kind, name, doc)
    end

File: test/support/with_without_module_doc.ex
Comment (josevalim) [reply test/support/with_without_module_doc.ex:1]: We don't need to test both autolink and ref test. We already test that autolink warns when something is private, we don't need to add new ones. My suggestion is to move this module to `ref_test` and remove the autolink test.

In a nutshell, autolink should only test the autolink and warning behaviour, it shouldn't really be testing what is public or private, because that's not its responsibility. If we are having a lot of tests around public and private already there, I recommend prunning/simplifying them. :)

File: lib/ex_doc/refs.ex
Comment (eksperimental) [reply lib/ex_doc/refs.ex:107]: I don't think we should have this visibility logic placed in a fetch function. Actually, the if i'm  not mistaken the very same bug that this PR solves is because of this logic being out of the visibility function.

File: lib/ex_doc/refs.ex
Comment (eksperimental) [reply lib/ex_doc/refs.ex:103]: True. fixed

File: test/support/with_without_module_doc.ex
Comment (eksperimental) [reply test/support/with_without_module_doc.ex]: Got you.
I cannot place the module in the .exs file, because that defeats the purpose of the module which is testing the visibility based on the @doc attributes.
I did my best to simplify the tests in refs_test.exs, please have a look

File: lib/ex_doc/refs.ex
Comment (josevalim) [reply lib/ex_doc/refs.ex]: You can put it inside the visibility function if you want. My point is that the logic today is more complicated than it needs to be, because you can work with the module_visibility you already got, instead of computing it over again.

Comment (josevalim) [reply lib/ex_doc/refs.ex]: But removing the `module` gets a similar result, so it looks good to me! :)

File: lib/ex_doc/refs.ex
Comment (josevalim) [inline lib/ex_doc/refs.ex:151]:
```suggestion
  defp visibility(_, {_kind, _name, :hidden}),
    do: :hidden
```

File: test/support/with_without_module_doc.ex
Comment (josevalim) [reply test/support/with_without_module_doc.ex:1]: The tests look great, thanks! Do we still need this @moduledoc now with these function refs? :)

**Review by josevalim — APPROVED** [review]

File: lib/ex_doc/refs.ex
Comment (eksperimental) [reply lib/ex_doc/refs.ex]: fixed

File: test/support/with_without_module_doc.ex
Comment (eksperimental) [reply test/support/with_without_module_doc.ex:1]: Fixed. Thanks

**Comment by eksperimental** [comment]
Thank you @jonatanklosko and @josevalim for your reviews

**Pull request merged**
